target_chains/sui/contracts/sources/merkle_tree.move

```diff
@@ -0,0 +1,276 @@
+        let bytes20 = bytes::take_bytes(&mut cursor, 20);
+        cursor::destroy_empty(cursor);
+        bytes20::from_bytes(bytes20)
+    }
+
+    fun emptyLeafHash(): Bytes20 {
+        let v = vector::empty();
+        bytes::push_u8(&mut v, MERKLE_EMPTY_LEAF_PREFIX);
+        hash(&v)
+    }
+
+    fun leafHash(data: &vector<u8>): Bytes20 {
+        bytes::push_u8(&mut *data, MERKLE_LEAF_PREFIX);
+        hash(data)
+    }
+
+    fun nodeHash(
+        childA: Bytes20,
+        childB: Bytes20
+    ): Bytes20 {
+        if (greaterThan(childA, childB)) {
+            (childA, childB) = (childB, childA);
+        };
+        let data_A = bytes20::data(&childA);
+        let data_B = bytes20::data(&childB);
+        vector::append(&mut data_A, data_B);
+        bytes::push_u8(&mut data_A, MERKLE_NODE_PREFIX);
+        hash(&data_A)
+    }
+
+    // greaterThan returns whether a is strictly greater than b
+    // note that data(&a) and data(&b) are both vectors of length 20
+    fun greaterThan(a: Bytes20, b: Bytes20): bool{
+        // aa and bb both have length 20
+        let aa = data(&a);
+        let bb = data(&b);
+        let i = 0;
+        while (i < 20){
+            if (*vector::borrow(&aa, i) > *vector::borrow(&bb, i)){
+                return true
+            } else if (*vector::borrow(&bb, i) > *vector::borrow(&aa, i)){
+                return false
+            };
+            i = i + 1;
+        };
+        false
+    }
+
+    // The Sui Move stdlb insert function shifts v[i] and subsequent elements to the right.
+    // We don't want this behavior, so we define our own vector insert function using swapping.
+    // Reference: https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/packages/move-stdlib/sources/vector.move
+    fun vectorInsert<T: drop>(a: &mut vector<T>, value: T, index: u64){
+        vector::push_back<T>(a, value); // push value to end
+        vector::swap_remove(a, index); // swap value to correct position and pop last value
+    }
+
+    // isProofValid checks whether a merkle proof is valid and returns (valid, offset)
+    fun isProofValid(
+        encodedProof: &mut Cursor<u8>,
+        root: Bytes20,
+        leafData: vector<u8>,
+    ): bool {
+
+        let currentDigest: Bytes20 = leafHash(&leafData);
+        let proofSize: u8 = deserialize::deserialize_u8(encodedProof);
+        let i: u8 = 0;
+        while (i < proofSize){
+            let siblingDigest: Bytes20 = bytes20::new(
+                deserialize::deserialize_vector(encodedProof, 20)
+            );
+
+            currentDigest = nodeHash(
+                currentDigest,
+                siblingDigest
+            );
+            i = i + 1;
+        };
+        bytes20::data(&currentDigest) == bytes20::data(&root)
+    }
+
+    // constructProofs constructs a merkle tree and returns the root of the tree as
+    // a Bytes20 as well as the vector of encoded proofs
+    fun constructProofs(
```

Comment:
per my comment in slack, i think it would be smart to generate a bunch of valid / invalid messages from the official rust implementation for these tests. That would ensure that some of the details around the hash truncation etc. are correct.

Reply:
I can do this in a later PR after those messages are generated